benchmarks/distributed/ddp/compare/python_ddp.py

```diff
@@ -0,0 +1,228 @@
+            self._MAX_BUFFER_SIZE = max_buffer_size
+
+        def __str__(self):
+            return "Bucket: num_params={}, total_elements={}, ready_param_grad_count={}".format(
+                len(self.param_to_offset),
+                self.total_elements,
+                self.ready_param_grad_count)
+
+        def is_full(self):
+            """
+            Returns whether grad for all the params in current bucket are ready
+            and copied to self.buffer.
+            """
+            assert self.ready_param_grad_count >= 0
+            assert self.ready_param_grad_count <= len(self.param_to_offset)
+            return len(self.param_to_offset) == self.ready_param_grad_count
+
+        def empty(self):
+            self.ready_param_grad_count = 0
+
+        def try_hold_param(self, param):
+            """
+            Checks whether current bucket has enough buffer to hold the incoming
+            param. If there is enough space, distribute param into current
+            bucket and Returns true. Otherwise, returns False.
+            """
+            if self.total_elements + param.numel() <= self._MAX_BUFFER_SIZE :
+                self.param_to_offset[param] = self.total_elements
+                self.total_elements += param.numel()
+                return True
+            else:
+                return False
+
+    def __init__(self, module, process_group, async_reduction=True, buffer_size=2 ** 22):
+        super(PythonDDP, self).__init__()
+
+        self.module = module
+        self.process_group = process_group
+        self.world_size = dist.get_world_size(group=self.process_group)
+        self.async_reduction = async_reduction
+
+        # Holds all_reduce handles, used when async_reduction is True
+        self.async_handles = set()
+
+        # Ensure buffer_size is large enough to hold largest param.
+        max_numel = max(p.numel() for p in module.parameters())
+        assert buffer_size > max_numel, "buffer_size: {} should be larger than largest param: {}".format(buffer_size, max_numel)
+
+        # Build buckets for params
+        self.param_to_bucket, self.buckets = self._build_buckets_for_params(buffer_size)
+
+        # Register per-parameter hook to be invoked when grad is ready.
+        for p in self.module.parameters():
+            assert p.requires_grad
+            p.register_hook(functools.partial(self._on_param_grad_ready, p))
+
+    def _build_buckets_for_params(self, max_buffer_size):
+        """
+        Distributes params into list of buckets. Maintains param -> bucket
+        mapping. Returns tuple of (param_to_buckets, buckets).
+        """
+        print("_build_buckets_for_params called")
+        params_to_buckets = {}
+        buckets = set()
+        cur_bucket = self.Bucket(max_buffer_size)
+        total_param = 0
+        for param in self.module.parameters():
+            total_param += 1
+            assert param.requires_grad, "param.requires_grad must be True"
```

Comment:
Do we have a lot of parameters/layers? If so, `assert` here can slow down the loop, as it requires a sync between host and device.

Reply:
there are 1M params. Good idea -- we can introduce debug level to toggle it.